File: 13-05-2024_Number_of_good_component.py

from typing import List

class Solution:
    def dfs(self, adj: List[List[int]], i: int, visited: List[bool]) -> None:
        visited[i] = True
        self.v_cnt += 1
        self.e_cnt += len(adj[i])

        for next_node in adj[i]:
            if not visited[next_node]:
                self.dfs(adj, next_node, visited)

    def findNumberOfGoodComponent(self, e: int, v: int, edges: List[List[int]]) -> int:
        visited = [False] * (v + 1)
        adj = [[] for _ in range(v + 1)]

        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])

        ans = 0

        for i in range(1, v + 1):
            self.v_cnt = 0
            self.e_cnt = 0
            if not visited[i]:
                visited[i] = True
                self.dfs(adj, i, visited)

                if self.e_cnt == self.v_cnt * (self.v_cnt - 1):
                    ans += 1

        return ans

File: 13-05-2024_Number_of_good_component.java
class Solution {
    static int v_cnt;
    static int e_cnt;

    public static void dfs(List<List<Integer>> adj, int i, boolean[] visited) {
        visited[i] = true;
        v_cnt++;
        e_cnt += adj.get(i).size();

        for (int next : adj.get(i)) {
            if (!visited[next]) {
                dfs(adj, next, visited);
            }
        }
    }

    public static int findNumberOfGoodComponent(int e, int v, int[][] edges) {
        List<List<Integer>> adj = new ArrayList<>();
        for (int i = 0; i <= v; i++) {
            adj.add(new ArrayList<>());
        }

        boolean[] visited = new boolean[v + 1];
        int ans = 0;

        for (int[] edge : edges) {
            adj.get(edge[0]).add(edge[1]);
            adj.get(edge[1]).add(edge[0]);
        }

        for (int i = 1; i <= v; i++) {
            v_cnt = 0;
            e_cnt = 0;
            if (!visited[i]) {
                visited[i] = true;
                dfs(adj, i, visited);

                if (e_cnt == (v_cnt * (v_cnt - 1))) {
                    ans++;
                }
            }
        }
        return ans;
    }
}
